src/libtcod/renderer_xterm.c

```diff
@@ -159,7 +174,125 @@ static void xterm_destructor(struct TCOD_Context* __restrict self) {
   xterm_cleanup();
 }
 
-TCOD_Context* TCOD_renderer_init_xterm(const char* window_title) {
+static void xterm_handle_input_escape() {
+  if (getchar() != '[')
+    return;
+  char rows_buf[16], cols_buf[16];
+  rows_buf[15] = '\0';
+  cols_buf[15] = '\0';
+  for (int i = 0; i < 15; i++) {
+    int ch = getchar();
+    if (ch == ';') {
+      rows_buf[i] = '\0';
+      break;
+    }
+    if (!isdigit(ch)) {
+      return;
+    }
+    rows_buf[i] = ch;
+  }
+  for (int j = 0; j < 15; j++) {
+    int ch = getchar();
+    if (ch == 'R') {
+      cols_buf[j] = '\0';
+      break;
+    }
+    if (!isdigit(ch)) {
+      return;
+    }
+    cols_buf[j] = ch;
+  }
+  g_got_rows = atoi(rows_buf);
+  g_got_columns = atoi(cols_buf);
+  g_got_size_timestamp = SDL_GetTicks();
+}
+
+static void *xterm_handle_input(void *arg) {
+  while (true) {
+    int ch = getchar();
+    if (ch == '\x1b') {
+      xterm_handle_input_escape();
+      continue;
+    }
+    int sym = ch;
+    int mod = KMOD_NONE;
+    if isupper(ch) {
+      sym = tolower(sym);
+      mod = KMOD_SHIFT;
+    }
+    SDL_Event down_event;
+    down_event.type = SDL_KEYDOWN;
+    down_event.key.timestamp = SDL_GetTicks();
+    down_event.key.windowID = 0;
+    down_event.key.state = SDL_PRESSED;
+    down_event.key.repeat = 0;
+    down_event.key.keysym.sym = sym;
+    down_event.key.keysym.scancode = SDL_GetScancodeFromKey(sym);
+    down_event.key.keysym.mod = mod;
+    SDL_Event text_event;
+    text_event.type = SDL_TEXTINPUT;
+    text_event.text.timestamp = SDL_GetTicks();
+    text_event.text.windowID = 0;
+    text_event.text.text[0] = ch;
+    text_event.text.text[1] = '\0';
+    SDL_Event up_event = down_event;
+    up_event.type = SDL_KEYUP;
+    up_event.key.state = SDL_RELEASED;
+    SDL_PushEvent(&down_event);
+    SDL_PushEvent(&text_event);
+    SDL_PushEvent(&up_event);
+  }
+  return NULL;
+}
+
+static TCOD_Error xterm_recommended_console_size(
+    struct TCOD_Context* __restrict self,
+    float magnification,
+    int* __restrict columns,
+    int* __restrict rows) {
+  fprintf(stdout, "\x1b[%i;%iH", INT_MAX, INT_MAX);
+  fflush(stdout);
+  Uint32 start_time = SDL_GetTicks();
+  fprintf(stdout, "\x1b[6n");
+  fflush(stdout);
+  while (SDL_GetTicks() < start_time + 100) {
```

Comment:
Use [SDL_TICKS_PASSED](https://wiki.libsdl.org/SDL_TICKS_PASSED) to avoid overflow issues.
```suggestion
  while (!SDL_TICKS_PASSED(SDL_GetTicks(), start_time + 100)) {
```